keps/sig-node/3063-dynamic-resource-allocation/README.md

```diff
@@ -1963,6 +1964,40 @@ gRPC interface. It was inspired by
 [CSI](https://github.com/container-storage-interface/spec/blob/master/spec.md),
 with “volume” replaced by “resource” and volume specific parts removed.
 
+##### `WatchConfiguration`
+This RPC is called by the kubelet just after a Plugin gets registered.
+
+```protobuf
+rpc WatchConfiguration(ConfigurationRequest) returns (stream ConfigurationResponse) {}
+```
+
+It sends a `ConfigurationRequest` message and receives a stream of
+`ConfigurationResponse` messages corresponding to dynamic changes in the
+configuration of the Plugin.
+
+```protobuf
+message ConfigurationRequest {}
+
+message ConfigurationResponse {
+    // The list of driver names for which resources should be processed.
+    repeated string accepted_driver_names = 1;
+}
+```
+
+For now the `ConfigurationRequest` message is empty and the
+`ConfigurationResponse` only contains a single field for
+`accepted_driver_names`. Over time we imagine these messages will be expanded
+to accommodate more fields.
+
+The list of `accepted_driver_names` in the most recent
+`ConfigurationResponse` message corresponds to the list of driver names for
+which the Plugin should prepare and unprepare resources.
+
+That is, for each incoming `ResourceClaim` associated with a given driver name,
+the kubelet MUST call `NodePrepareResources` and `NodeUnprepareResources` for
+each Plugin that has that driver name in its current list of
+`accepted_driver_names`.
```

Comment:
Just double checking: this wording implies that more than one driver might have to be called for a claim ("for each Plugin that has..."). Is that intended?

Reply:
What happens if `NodePrepareResources` gets called for a driver, the driver removes the driver name for that claim from `accepted_driver_names`, and then `NodeUnprepareResources` needs to be called?

Reply:
> Just double checking: this wording implies that more than one driver might have to be called for a claim ("for each Plugin that has..."). Is that intended?

Yes, and in fact part of what this is designed for -- a single claim being used to co-allocate multiple resources. Please see https://github.com/kubernetes/kubernetes/issues/115263#issuecomment-1401862480 for one concrete use case.

> What happens if `NodePrepareResources` gets called for a driver, the driver removes the driver name for that claim from `accepted_driver_names`, and then `NodeUnprepareResources` needs to be called?

When I was putting this together, I thought about that case, and concluded that it would be similar to the plugin itself temporarily going ofline (meaning that the kubelet would not be able to call `NodeUnprepareResources` on the "default" driver name associated with the plugin).

However, thinking about it again now -- it's actually not the same. If the plugin is offline, then the `NodeUnprepareResources` call against its default driver name will fail, and be retried once the plugin comes back online. This is not true for the list of configured `accepted_driver_names` though.

There could be issues if care is not taken to ensure that `NodeUnprepareResources` is called (and succeeds) against *all*  kubelet plugins where `NodePrepareResources` was called for a given claim.

I will need to think about how best to ensure this. I don't think we want to be tracking state about every outstanding claim inside the kubelet (which would be the obvious way to deal with this).

Reply:
This scenario has very low probability to happen, but it could. However, in any implementation if plugin is offline for certain time it will not be called by Kubelet just because it's offline or because Kubelet tried to call it several times and failed.

Reply:
I disagree. The difference here is that nothing will ever fail (the kubelet will never know it's supposed to). When we have a 1:1 mapping of driver name to plugin name, then the kubelet knows a plugin is missing and it will continue to fail the NodeUnrepareResource call until the plugin comes back.

In the scenario described above, the kubelet never knows that the plugin was supposed to be invoked in the first place and has no opportunity to fail if it is not present (it silently succeeds) and it appears to the system as if the claim has been successfully deallocated (when it hasn't).

Reply:
A different alternative (which I had proposed earlier on, but was rejected in favor of not changing the core API) would be to reverse things.That is, not have this API at all, and instead include the list of plugin names that should be called out to as part of the `ResourceClaimStatus`.

Right now we have a single field for `DriverName` in there, but we could recast that as the name of the controller and add an additional field to list out the names of the plugins to be called. These could then be populated on a claim-by-claim basis by the controller once it knows which resources it is actually reserving as part of a claim allocation.

Reply:
I like changing ResourceClaimsStatus better. It's more logical and transparent approach than hacking it on a Kubelet level in my opinion.

Reply:
@pohly What do you think?

Would you be willing to consider an extension to the `ResourceClaimStatus` with a field called `NodePlugins` (naming up for debate)?

Where the semantics in the kubelet that interpret this are:
1. If `NodePlugins` is empty, use `DriverName` (for backwards compatibility)
2. Else, call out to all `NodePlugins` for this claim

And then we extend the controller library to allow drivers to populate this from a new field in the `AllocationResult`.

Reply:
Adding such a `NodePlugins` field makes sense to me.

Should we support per-plugin `ResourceHandle`, too? I think that would make sense: consider a "meta controller" which fronts several independent third-party drivers. If those are truly independent and at least two of them have their own notion of what needs to be in `ResourceHandle`, then we need to store these values separately to enable kubelet to pass the right one to each node plugin.

> Where, if they don't set it, the library sets NodePlugins = []string{DriverName}.

We can (should?) leave it empty in this case because that is defined as the fallback.

Reply:
A per-plugin `ResourceHandle` would probably make sense.

Does this imply that maybe instead of a `NodePlugins` field, we actually want to turn  `ResourceClaimStatus.Allocation` into a list of some sort, i.e.

Either this:
```
type ResourceClaimStatus struct {
	DriverName string
	Allocation map[string]AllocationResult
	ReservedFor []ResourceClaimConsumerReference
	DeallocationRequested bool
}
```
Where the `string` in `map[string]AllocationResult` corresponds to a plugin name

Or this:
```
type ResourceClaimStatus struct {
	DriverName string
	Allocation []AllocationResult
	ReservedFor []ResourceClaimConsumerReference
	DeallocationRequested bool
}

// AllocationResult contains attributed of an allocated resource.
type AllocationResult struct {
    KubeletPluginName string
    ...
}
```
Where the name of the kubelet plugin is now explicitly embedded in the `AllocationResult`.

I guess either way we are looking at a bump to `v1alpha2` to support this.

Reply:
The second option with `AllocationResult` is what we should go for. Literal `map` types are frowned upon for Kubernetes APIs. Instead one should use a slice and `+listType=map, +listMapKey=kubeletPluginName`.

Reply:
I'll update this pull request with the new proposed changes.

Reply:
Updated.